src/dev/lupio/lupio_blk.cc

```diff
@@ -0,0 +1,233 @@
+{
+    return val ? __builtin_clzll(val) : 64;
+}
+
+void
+LupioBLK::dmaEventDone()
+{
+    if (writeOp) {
+        for (int i = 0; i < reqLen; i += SECTOR_SIZE){
+            image.write(&reqData[i], offset);
+            offset++;
+        }
+    }
+    delete[] reqData;
+    DPRINTF(LupioBLK, "Done with DMA event\n");
+    platform.postPciInt(LUPIO_BLK_INT_ID);
+}
+
+uint32_t
+LupioBLK::lupioBLKRead(const uint8_t addr)
+{
+    uint32_t r = 0;
+
+    switch (addr >> 2) {
+        case LUPIO_BLK_CONF:
+            r = SECTOR_BITS << 16               /* log2(512B / block) */
+                | (63 - clz64(nbBlocks));      /* log2(Nb of blocks) */
+
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_CONF: %d\n", r);
+            break;
+        case LUPIO_BLK_NBLK:
+            r = nblk;
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_NBLK: %d\n", r);
+            break;
+        case LUPIO_BLK_BLKA:
+            r = lba;
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_BLKA: %d\n", r);
+            break;
+        case LUPIO_BLK_MEMA:
+            r = mem;
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_MEMA: %d\n", r);
+            break;
+
+        case LUPIO_BLK_STAT:
+            r = 0;                      /* Always idle */
+            if (writeOp) {
+                r |= LUPIO_BLK_TYPE;    /* Write command */
+            }
+            if (err) {
+                r |= LUPIO_BLK_ERRR;    /* Error */
+            }
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_STAT: %d\n", r);
+
+            /* Acknowledge IRQ */
+            platform.clearPciInt(LUPIO_BLK_INT_ID);
+            break;
+
+        default:
+            panic("Unexpected read to the LupioRTC device at address %#llx!",
+                    addr);
+            break;
+    }
+    return r;
+}
+
+void
+LupioBLK::lupioBLKWrite(const uint8_t addr, uint32_t val)
+{
+    switch (addr >> 2) {
+        case LUPIO_BLK_NBLK:
+            nblk = val;
+            DPRINTF(LupioBLK, "Write LUPIO_BLK_NBLK: %d\n", nblk);
+            break;
+        case LUPIO_BLK_BLKA:
+            lba = val;
+            DPRINTF(LupioBLK, "Write LUPIO_BLK_BLKA: %d\n", lba);
+            break;
+        case LUPIO_BLK_MEMA:
+            mem = val;
+            DPRINTF(LupioBLK, "Write LUPIO_BLK_MEMA: %d\n", mem);
+            break;
+
+        case LUPIO_BLK_CTRL:
+            err = false;
+            writeOp = val & LUPIO_BLK_TYPE;
+
+            /* Perform command */
+            lupioBLKCmd();
+
+            break;
+
+        default:
+            panic("Unexpected write to the LupioRTC device at address %d!",
+                    addr);
+            break;
+    }
+}
+
+
+void
+LupioBLK::lupioBLKCmd(void)
+{
+    // Check parameters
+    if ((uint64_t)lba + nblk > nbBlocks) {
+        panic("Bad LBA range\n");
+    }
+
+    // Perform transfer
+    reqLen = nblk * SECTOR_SIZE;
+    DPRINTF(LupioBLK, "Write op start length: %d\n", reqLen);
+    reqData = new uint8_t[reqLen];
+    offset = lba;
```

Comment:
`offset` should simply be a local variable, both here and in DmaDone(). There's no need to make it a class member.